src/tests/ftest/util/yaml_utils.py

```diff
@@ -0,0 +1,485 @@
+
+    Returns:
+        list: A list of detected devices - empty if none found
+
+    """
+    log = getLogger()
+    devices = []
+
+    # Find the devices on each host
+    if device_type == "VMD":
+        # Exclude the controller revision as this causes heterogeneous clush output
+        command_list = [
+            "/sbin/lspci -D",
+            "grep -E '^[0-9a-f]{{{0}}}:[0-9a-f]{{2}}:[0-9a-f]{{2}}.[0-9a-f] '".format(length),
+            "grep -E 'Volume Management Device NVMe RAID Controller'",
+            r"sed -E 's/\(rev\s+([a-f0-9])+\)//I'"]
+    elif device_type == "NVMe":
+        command_list = [
+            "/sbin/lspci -D",
+            "grep -E '^[0-9a-f]{{{0}}}:[0-9a-f]{{2}}:[0-9a-f]{{2}}.[0-9a-f] '".format(length),
+            "grep -E 'Non-Volatile memory controller:'"]
+        if device_filter and device_filter.startswith("-"):
+            command_list.append("grep -v '{}'".format(device_filter[1:]))
+        elif device_filter:
+            command_list.append("grep '{}'".format(device_filter))
+    else:
+        raise CommandFailure(
+            "Invalid 'device_type' for NVMe/VMD auto-detection: {}".format(device_type))
+
+    command = " | ".join(command_list) + " || :"
+    task = run_task(host_list, command, verbose=True)
+
+    # Verify the command was successful on each server host
+    if display_task(task):
+        # Verify each server host has the same VMD PCI addresses
+        output_data = list(task.iter_buffers())
+        if len(output_data) > 1:
+            log.error("ERROR: Non-homogeneous {} PCI addresses.".format(device_type))
+        elif len(output_data) == 1:
+            # Get the devices from the successful, homogeneous command output
+            output_str = "\n".join([line.decode("utf-8") for line in output_data[0][0]])
+            devices = find_pci_address(output_str)
+    else:
+        log.error("ERROR: Issuing command '{}'".format(command))
+
+    return devices
+
+
+def get_vmd_address_backed_nvme(host_list, vmd_disks, vmd_controllers):
+    """Find valid VMD address which has backing NVMe.
+
+    Args:
+        host_list (NodeSet): hosts on which to find the VMD addresses
+        vmd_disks (list): list of PCI domain numbers for each VMD controlled disk
+        vmd_controllers (list): list of PCI domain numbers for each VMD controller
+
+    Returns:
+        list: a list of the VMD controller PCI domain numbers which are connected to the VMD disks
+
+    """
+    log = getLogger()
+    disk_controllers = []
+    command_list = ["ls -l /sys/block/", "grep nvme"]
+    if vmd_disks:
+        command_list.append("grep -E '({0})'".format("|".join(vmd_disks)))
+    command_list.extend(["cut -d'>' -f2", "cut -d'/' -f4"])
+    command = " | ".join(command_list) + " || :"
+    task = run_task(host_list, command, verbose=True)
+
+    # Verify the command was successful on each server host
+    if not display_task(task):
+        log.error("ERROR: Issuing command '{}'".format(command))
```

Comment:
(pylint-logging-format-interpolation) Use lazy % formatting in logging functions